jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DefaultServletCombinationsTest.java

```diff
@@ -0,0 +1,191 @@
+        FS.ensureDirExists(docRoot);
+        Path subdirPath = docRoot.resolve("subdir");
+        FS.ensureDirExists(subdirPath);
+        Files.writeString(docRoot.resolve("index.html"), "Static index.html at root", UTF_8);
+        Files.writeString(docRoot.resolve("foo.welcome"), "Static foo.welcome at root", UTF_8);
+        Files.writeString(subdirPath.resolve("index.html"), "Static index.html at subdir", UTF_8);
+        Files.writeString(subdirPath.resolve("foo.welcome"), "Static foo.welcome at subdir", UTF_8);
+
+        server = new Server();
+
+        connector = new LocalConnector(server);
+        connector.getConnectionFactory(HttpConfiguration.ConnectionFactory.class).getHttpConfiguration().setSendServerVersion(false);
+
+        context = new ServletContextHandler();
+        context.setBaseResourceAsPath(docRoot);
+        context.setContextPath("/");
+        context.setWelcomeFiles(new String[]{"index.html", "index.welcome"});
+
+        ServletHolder welcomeExtHolder = context.addServlet(WelcomeServlet.class, "*.welcome");
+        welcomeExtHolder.setInitParameter("mapping", "welcome extension");
+    }
+
+    private void startServer(boolean pathInfoOnly, ResourceService.WelcomeMode welcomeMode) throws Exception
+    {
+        ServletHolder defaultHolder = context.addServlet(DefaultServlet.class, "/");
+        defaultHolder.setInitParameter("pathInfoOnly", String.valueOf(pathInfoOnly));
+
+        server.setHandler(context);
+        server.addConnector(connector);
+
+        server.start();
+        ((DefaultServlet)defaultHolder.getServlet()).getResourceService().setWelcomeMode(welcomeMode);
+    }
+
+    @AfterEach
+    public void destroy() throws Exception
+    {
+        server.stop();
+        server.join();
+    }
+
+    public static class WelcomeServlet extends HttpServlet
+    {
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
+        {
+            resp.setStatus(200);
+            String mapping = getInitParameter("mapping");
+            resp.getWriter().print("Servlet at " + mapping);
+        }
+    }
+
+    record Data(boolean pathInfoOnly, ResourceService.WelcomeMode welcomeMode, String requestPath, int expectedStatus, String expected)
+    {
+    }
+
+    public static Stream<Data> data()
+    {
+        return Stream.of(
+            new Data(false, SERVE, "/", HttpStatus.OK_200, "Static index.html at root"),
+            new Data(false, REDIRECT, "/", HttpStatus.FOUND_302, "http://local/index.html"),
+            new Data(false, REHANDLE, "/", HttpStatus.OK_200, "Static index.html at root"),
+            new Data(true, SERVE, "/", HttpStatus.FOUND_302, "http://local/index.html/"),
+            new Data(true, REDIRECT, "/", HttpStatus.FOUND_302, "http://local/index.html"),
+            new Data(true, REHANDLE, "/", HttpStatus.FOUND_302, "http://local/index.html/"),
```

Comment:
```suggestion
            new Data(true, SERVE, "/static/", HttpStatus.FOUND_302, "http://local/index.html/"),
            new Data(true, REDIRECT, "/static/", HttpStatus.FOUND_302, "http://local/index.html"),
            new Data(true, REHANDLE, "/static/", HttpStatus.FOUND_302, "http://local/index.html/"),
```

You can do that explicitly here or perhaps prefix the "/static" in the Data constructor if pathInfoOnly is true?